ykpack/src/types.rs

```diff
@@ -195,6 +197,7 @@ pub enum Rvalue {
     BinaryOp(BinOp, Operand, Operand),
     CheckedBinaryOp(BinOp, Operand, Operand),
     Unimplemented,
+    Load(LocalIndex), // FIXME arg may not be local.
```

Comment:
Perhaps we should, at least temporarily, call this `LoadLocal`?

Reply:
This will be fixed very soon, so I wouldn't worry about it for now.

Reply:
The immediate next PR or ...?

Reply:
It'll be either the next or the one after.

I don't want to start fleshing out instructions implementations until I know that I can deal with other operand types. Spent some time the other day looking at how constants are born and I have a plan.

Reply:
Then I would still prefer to call this `LoadLocal` and then we no longer have the `FIXME` and the wait for a couple of PRs. As you can tell, I am allergic to `FIXME`s :p

Reply:
We can call it `LoadLocal` if you wan't, but I'd still want a `FIXME` saying "this shouldn't be `LoadLocal`, it should be `Load` and accept other operand types too" :P

I suspect we are worrying too much too early. I don't even know if this approach is going to work yet.

Reply:
Having thought some more: here's what we should do for the *next* PR.

Keep this statement as `Load`, but change the `Local` inside to `Operand` and have `Operand` be an enum with (for now) one `Local` variant. This implies there are more to come.

It'll be quite a large change because we already have a `Operand` enum which contains a load of old stuff and tests rely on them. I'd planned to sort this later, but we can bring that work forward to keep everyone happy :)

Sound good?

Reply:
We don't even really need to make a decision on this yet: if we say `LoadLocal` now, in a future PR, we might add `LoadGlobal` or rename this to `Load` and use an internal `enum` or whatever (FWIW, it's probably easier to have separate variants for LoadLocal/LoadGlobal rather than an internal VM, but this is partly a matter of taste). The beauty of calling it `LoadLocal` is that it doesn't need a `FIXME`: the name is impossible to misread.

Reply:
Wait, there seems to be confusion over the role of the term "Local" here.

LLVM's `load` instruction loads something from memory. Where to load from is expressed by a pointer operand:
https://llvm.org/docs/LangRef.html#load-instruction

My `Load(Local)` variant allows a load from memory where the pointer operand is stored in a local variable. i.e.
```
%p = <some-pointer>
%r = load(%p)
```

Adding new instructions for different operand types doesn't seem like the right thing to do. Consider what would happen to a `Add` and `Sub`:
```
enum Statement {
    ...
    AddConstConst(Const, Const)
    AddLocalLocal(Local, Local)
    AddConstGlobal(Const, Global)
    AddLocalGlobal(Local, Global)
    AddGlobalGlobal(Global, Global)
    AddLocalConst(Local, Const)
    SubConstConst(Const, Const)
    SubLocalLocal(Local, Local)
    SubConstGlobal(Const, Global)
    SubLocalGlobal(Local, Global)
    SubGlobalGlobal(Global, Global)
    SubLocalConst(Local, Const)
}
```

It'd better be expressed as:
```
enum Statement {
    ...
    Add(Operand, Operand)
    Sub(Operand, Operand)
}

enum Operand {
    Local,
    Global,
    Const
}
```

Makes sense? Sorry, it's quite hard to explain.

Reply:
Dumb question: do we have to follow LLVM slavishly here or is this our own IR where we can structure things as we see fit?

Reply:
We have a little bit of wiggle room, but not a lot. Sadly our IR is going to have to be very similar to LLVM's due to the way the codegen works.

Reply:
Yes, I understand that we're currently somewhat constrained, but does it need to be exactly the same in this case?

Reply:
It may help to look at the point where the [generic backend code calls the llvm backend for a load](https://github.com/softdevteam/ykrustc/blob/043bf15b2c4fd0b21c02f6fe3d82bb8f663faf75/src/librustc_codegen_llvm/builder.rs#L438-L444):

```
    fn load(&mut self, ptr: &'ll Value, align: Align) -> &'ll Value {
        unsafe {
            let load = llvm::LLVMBuildLoad(self.llbuilder, ptr, UNNAMED);
            llvm::LLVMSetAlignment(load, align.bytes() as c_uint);
            load
        }
    }
```

`codegen_ssa` calls this passing a C++ pointer (`ptr`) which represents where to load from and returns another C++ pointer representing the result of the load.

In order for the Sir generator to know what `ptr` is, we must have cached its creation during an earlier call to the LLVM API.

That's the granularity we are working at unfortunately. We have to decide how each individual call to the LLVM API should mutate the SIR.

You are wondering why we are working so close to LLVM?

Reply:
> You are wondering why we are working so close to LLVM?

Yes, partly.